libimage/runtime.go

```diff
@@ -608,11 +616,37 @@ func (r *Runtime) ListImages(ctx context.Context, names []string, options *ListI
 	// as the layer tree will computed once for all instead of once for
 	// each individual image (see containers/podman/issues/17828).
 
-	tree, err := r.layerTree(ctx, images)
+	var tree *layerTree
+	getTree := func() (*layerTree, error) {
+		if tree == nil {
+			if options.SetListData {
+				tree, err = r.layerTree(images, snapshot.Layers)
+			} else {
+				tree, err = r.layerTree(nil, nil)
```

Comment:
This is really preexisting, but also not great. We can return data which is inconsistent with the filter conditions.